File: dotnet/src/Envdoctor/Core/Audit.cs
using Envdoctor.Config;
using Envdoctor.Detectors;
using Envdoctor.Models;
using Envdoctor.Utils;

namespace Envdoctor.Core;

public static class Audit
{
    /// Detector ids in a well-known order. The order here is the order
    /// findings are produced.
    public static List<IDetector> AllDetectors() =>
        new()
        {
            new MissingDetector(),
            new UnusedDetector(),
            new UndefinedSourceDetector(),
            new DuplicatesDetector(),
            new EnvironmentDiffDetector(),
            new TypeMismatchDetector(),
            new PublicPrefixDetector(),
            new WeakSecretDetector(),
            new TypoDetector(),
            new SchemaValidationDetector(),
        };

    /// Run the full audit pipeline over a model and return aggregated findings.
    public static List<Finding> RunAudit(ProjectModel model, EnvdoctorConfig config, IndexedModel index)
    {
        var findings = new List<Finding>();
        foreach (var detector in AllDetectors())
            findings.AddRange(detector.Detect(index));

        findings = ApplyRuleSeverities(findings, config);
        findings = ApplyIgnores(findings, config);

        return findings;
    }

    /// Apply user-configured severity overrides for each rule. "off" removes
    /// the finding entirely, "error"/"warning" rewrite its severity.
    private static List<Finding> ApplyRuleSeverities(List<Finding> findings, EnvdoctorConfig config)
    {
        var result = new List<Finding>();
        foreach (var f in findings)
        {
            if (!config.Rules.TryGetValue(f.RuleId, out var overrideSev))
            {
                result.Add(f);
                continue;
            }
            switch (overrideSev)
            {
                case RuleSeverity.Off:
                    break;
                case RuleSeverity.Error:
                    var ef = f.Clone();
                    ef.Severity = Severity.Error;
                    result.Add(ef);
                    break;
                case RuleSeverity.Warning:
                    var wf = f.Clone();
                    wf.Severity = Severity.Warning;
                    result.Add(wf);
                    break;
            }
        }
        return result;
    }

    /// Drop findings whose variable matches any `ignoreVariables` glob.
    private static List<Finding> ApplyIgnores(List<Finding> findings, EnvdoctorConfig config)
    {
        if (config.IgnoreVariables.Count == 0)
            return findings;
        return findings.Where(f => !Glob.MatchesAnyGlob(config.IgnoreVariables, f.Variable)).ToList();
    }
}

File: swift/Sources/EnvdoctorKit/Core/Audit.swift
import Foundation

public struct AuditOptions {
    public var strict: Bool
    /// Restrict the audit to specific detector ids (e.g. `--only missing`).
    public var only: [String]
    /// Per-detector severity overrides.
    public var rules: [String: RuleSeverity]

    public init(strict: Bool, only: [String] = [], rules: [String: RuleSeverity] = [:]) {
        self.strict = strict
        self.only = only
        self.rules = rules
    }
}

/// The audit engine is fully format-agnostic: it takes a `ProjectModel` and
/// returns an `AuditResult`. Detectors run in a stable order and their
/// findings are aggregated; the exit code is derived from severity (and
/// `strict`).
public enum Audit {
    /// All detectors, in a stable order. The audit engine runs them in this
    /// order and the report renders in the same order.
    public static let allDetectors: [Detector] = [
        MissingDetector(),
        UnusedDetector(),
        UndefinedSourceDetector(),
        DuplicatesDetector(),
        EnvironmentDiffDetector(),
        TypeMismatchDetector(),
        PublicPrefixDetector(),
        WeakSecretDetector(),
        TypoDetector(),
        SchemaValidationDetector(),
    ]

    public static func runAudit(_ model: ProjectModel, options: AuditOptions) -> AuditResult {
        let index = buildIndex(model)

        let detectors = options.only.isEmpty
            ? allDetectors
            : allDetectors.filter { options.only.contains($0.id) }

        var findings = detectors.flatMap { $0.detect(index) }

        findings = applyIgnores(findings, model)
        findings = applyRuleSeverities(findings, options.rules)

        let summary = AuditSummary(
            filesScanned: model.allFiles.count,
            variablesFound: Pipeline.distinctVariableCount(model),
            errors: findings.filter { $0.severity == .error }.count,
            warnings: findings.filter { $0.severity == .warning }.count,
            infos: findings.filter { $0.severity == .info }.count,
            total: findings.count
        )

        let exitCode = ExitCodes.auditExitCode(ExitContext(findings: findings, strict: options.strict))
        return AuditResult(findings: findings, summary: summary, exitCode: exitCode)
    }

    /// Build a map of variable name → set of ignored rule ids from inline comments.
    private static func buildIgnoredRulesByName(_ model: ProjectModel) -> [String: Set<String>] {
        var ignored: [String: Set<String>] = [:]
        for file in model.envFiles {
            if file.environment == "example" { continue }
            for v in file.variables {
                guard let ignoreRules = v.ignoreRules, !ignoreRules.isEmpty else { continue }
                ignored[v.name, default: []].formUnion(ignoreRules)
            }
        }
        return ignored
    }

    /// Drop findings that are ignored inline in env files.
    private static func applyIgnores(_ findings: [Finding], _ model: ProjectModel) -> [Finding] {
        let ignored = buildIgnoredRulesByName(model)
        return findings.filter { f in
            guard let rules = ignored[f.variable] else { return true }
            return !rules.contains(f.ruleId)
        }
    }

    /// Apply per-detector severity overrides from config. Disabled rules are dropped.
    private static func applyRuleSeverities(_ findings: [Finding], _ rules: [String: RuleSeverity]) -> [Finding] {
        var result: [Finding] = []
        for var f in findings {
            guard let override = rules[f.ruleId] else {
                result.append(f)
                continue
            }
            if override == .off { continue }
            switch override {
            case .error: f.severity = .error
            case .warning: f.severity = .warning
            case .off: continue
            }
            result.append(f)
        }
        return result
    }

    /// True when a name is defined (has a value) in at least one environment file.
    public static func isDefinedInAnyEnv(_ model: ProjectModel, _ name: String) -> Bool {
        model.envFiles.contains { file in
            file.variables.contains { $0.name == name }
        }
    }

    /// True when a name is documented in `.env.example`.
    public static func isDocumented(_ model: ProjectModel, _ name: String) -> Bool {
        model.envFiles.contains { file in
            file.environment == "example" && file.variables.contains { $0.name == name }
        }
    }
}
